hazelcast/src/main/java/com/hazelcast/query/impl/bitmap/BitmapAlgorithms.java

```diff
@@ -73,23 +84,51 @@ public long getIndex() {
         public long advance() {
             long current = index;
 
-            long currentMax = iterators[0].getIndex();
-            long max = AscendingLongIterator.END;
-            while (currentMax != max) {
-                max = currentMax;
-                for (AscendingLongIterator iterator : iterators) {
-                    currentMax = iterator.advanceAtLeastTo(currentMax);
-                    if (currentMax == AscendingLongIterator.END) {
-                        index = AscendingLongIterator.END;
-                        return current;
-                    }
+            Arrays.sort(nodes);
+
+            Node first = nodes[0];
+            long min = first.iterator.getIndex();
+            Node last = nodes[nodes.length - 1];
+            long max = last.iterator.getIndex();
+
+            if (min != max && min != AscendingLongIterator.END) {
+                // No match: build a linked list.
+
+                for (int i = 0; i < nodes.length - 1; ++i) {
+                    nodes[i].next = nodes[i + 1];
                 }
+                last.next = null;
             }
 
-            // make a progress
-            iterators[0].advance();
+            while (min != max && min != AscendingLongIterator.END) {
+                max = first.iterator.advanceAtLeastTo(max);
+                if (max == AscendingLongIterator.END) {
+                    // End reached at the maximum iterator.
+                    index = AscendingLongIterator.END;
+                    return current;
+                }
+
+                // The first iterator is now guaranteed to be at the maximum or
+                // beyond it, make it last and repeat.
+
+                Node newFirst = first.next;
+                first.next = null;
+                last.next = first;
+                last = first;
+                first = newFirst;
 
-            index = max;
+                // The new first iterator is now our minimum.
+                min = first.iterator.getIndex();
+            }
+
+            if (min != AscendingLongIterator.END) {
+                // All iterators are at the same index: advance.
+                for (Node node : nodes) {
+                    node.iterator.advance();
```

Comment:
Shall we advance all iterators here? It seems it is enough to advance only the last one.

Reply:
fixed in https://github.com/hazelcast/hazelcast/pull/16162/commits/a47bef74a9d1ba9923f0356a46b8a9463fcb4cd7, also have optimized `advanceAtLeastTo` a little.